File: stream_add.cpp
/*
 * stream_add.cpp — Stream 64 MB float vectors and time just the add step.
 *
 * Layout
 * ------
 *  VectorPair   – RAII struct holding two 64 MB float arrays
 *  StreamSource – input_iterator that yields VectorPairs one at a time
 *  timed_add()  – adds two vectors; timer wraps ONLY the add loop
 *  main()       – drives the stream, prints per-pair stats and a summary
 *
 * Build:
 *   g++ -O2 -std=c++17 -o stream_add stream_add.cpp
 *
 * The -O2 flag lets the compiler vectorise the add loop (AVX/SSE) without
 * removing it, so the timing reflects realistic hardware throughput.
 */

#include <chrono>
#include <cmath>
#include <iomanip>
#include <iostream>
#include <algorithm>
#include <iterator>
#include <numeric>
#include <random>
#include <string>
#include <vector>

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------
static constexpr std::size_t MB              = 1024ULL * 1024;
static constexpr std::size_t VECTOR_BYTES    = 64 * MB;              // 64 MB
static constexpr std::size_t ELEMENTS        = VECTOR_BYTES / sizeof(float); // 16 777 216
static constexpr int         NUM_PAIRS       = 8;

// ---------------------------------------------------------------------------
// 1. Data types
// ---------------------------------------------------------------------------
struct VectorPair {
    int                 index;
    std::vector<float>  a;
    std::vector<float>  b;
};

// ---------------------------------------------------------------------------
// 2. StreamSource — generates VectorPairs lazily (one at a time)
// ---------------------------------------------------------------------------
class StreamSource {
public:
    explicit StreamSource(int num_pairs, std::uint64_t seed = 42)
        : num_pairs_(num_pairs), rng_(seed)
    {
        advance();   // prime first pair
    }

    bool has_next() const { return current_index_ < num_pairs_; }

    VectorPair next() {
        VectorPair p = std::move(current_);
        ++current_index_;
        if (has_next()) advance();
        return p;
    }

private:
    void advance() {
        std::uniform_real_distribution<float> dist(0.0f, 1.0f);
        current_.index = current_index_;
        current_.a.resize(ELEMENTS);
        current_.b.resize(ELEMENTS);
        for (std::size_t i = 0; i < ELEMENTS; ++i) {
            current_.a[i] = dist(rng_);
            current_.b[i] = dist(rng_);
        }
    }

    int             num_pairs_;
    int             current_index_ = 0;
    std::mt19937_64 rng_;
    VectorPair      current_;
};

// ---------------------------------------------------------------------------
// 3. Timed add — isolates *just* the addition loop
// ---------------------------------------------------------------------------
struct AddResult {
    std::vector<float>  result;
    double              elapsed_sec;
};

AddResult timed_add(const std::vector<float>& a, const std::vector<float>& b)
{
    std::vector<float> result(a.size());

    // ── just the add ──────────────────────────────────────────────────────
    auto t0 = std::chrono::high_resolution_clock::now();

    for (std::size_t i = 0; i < a.size(); ++i)
        result[i] = a[i] + b[i];

    auto t1 = std::chrono::high_resolution_clock::now();
    // ──────────────────────────────────────────────────────────────────────

    double elapsed = std::chrono::duration<double>(t1 - t0).count();
    return {std::move(result), elapsed};
}

// ---------------------------------------------------------------------------
// 4. Runtime estimation helpers
// ---------------------------------------------------------------------------
double estimate_remaining(const std::vector<double>& times, int pairs_left)
{
    if (times.empty()) return std::numeric_limits<double>::quiet_NaN();
    double mean = std::accumulate(times.begin(), times.end(), 0.0) / times.size();
    return mean * pairs_left;
}

double throughput_gbs(double elapsed_sec)
{
    double bytes_touched = 3.0 * static_cast<double>(VECTOR_BYTES); // read a, read b, write result
    return (bytes_touched / elapsed_sec) / (1024.0 * 1024.0 * 1024.0);
}

// ---------------------------------------------------------------------------
// 5. Main
// ---------------------------------------------------------------------------
int main()
{
    auto sep62 = std::string(62, '=');
    auto dash62 = std::string(62, '-');

    std::cout << sep62 << "\n";
    std::cout << "  Stream-add benchmark (C++)\n";
    std::cout << "  Vector size : " << (VECTOR_BYTES / MB) << " MB"
              << "  (" << ELEMENTS << " float32 elements)\n";
    std::cout << "  Pairs       : " << NUM_PAIRS << "\n";
    std::cout << sep62 << "\n";

    StreamSource source(NUM_PAIRS);
    std::vector<double> add_times;
    add_times.reserve(NUM_PAIRS);

    while (source.has_next()) {
        VectorPair pair = source.next();
        int pairs_done = pair.index + 1;
        int pairs_left = NUM_PAIRS - pairs_done;

        // ── just the add ────────────────────────────────────────────────
        AddResult ar = timed_add(pair.a, pair.b);
        // ────────────────────────────────────────────────────────────────

        add_times.push_back(ar.elapsed_sec);
        double est = estimate_remaining(add_times, pairs_left);

        std::cout << std::fixed << std::setprecision(3);
        std::cout << "  pair " << std::setw(2) << pairs_done << "/" << NUM_PAIRS
                  << " | add = " << std::setw(7) << ar.elapsed_sec * 1000.0 << " ms"
                  << " | " << std::setw(5) << std::setprecision(2) << throughput_gbs(ar.elapsed_sec) << " GB/s"
                  << " | est. remaining add time = ";
        if (pairs_left > 0)
            std::cout << std::setprecision(1) << est * 1000.0 << " ms\n";
        else
            std::cout << "done\n";

        // Release result memory immediately (benchmarking, not accumulating)
        ar.result.clear();
        ar.result.shrink_to_fit();
    }

    double total   = std::accumulate(add_times.begin(), add_times.end(), 0.0);
    double mean    = total / add_times.size();
    double minimum = *std::min_element(add_times.begin(), add_times.end());
    double maximum = *std::max_element(add_times.begin(), add_times.end());

    std::cout << dash62 << "\n";
    std::cout << std::fixed << std::setprecision(3);
    std::cout << "  Total add time : " << std::setw(8) << total   * 1000.0 << " ms\n";
    std::cout << "  Mean  add time : " << std::setw(8) << mean    * 1000.0 << " ms"
              << "  (" << std::setprecision(2) << throughput_gbs(mean) << " GB/s avg)\n";
    std::cout << "  Min   add time : " << std::setw(8) << std::setprecision(3) << minimum * 1000.0 << " ms\n";
    std::cout << "  Max   add time : " << std::setw(8) << maximum * 1000.0 << " ms\n";
    std::cout << sep62 << "\n";

    return 0;
}

File: stream_add.py
"""
stream_add.py — Stream 64 MB vectors and estimate runtime of just the add step.

Layout
------
1. stream_vectors()   – generator that yields numpy float32 arrays, each exactly 64 MB
2. timed_add()        – adds two vectors and returns (result, elapsed_seconds)
3. main()             – drives the stream, accumulates timing stats, prints a report
"""

import time
import numpy as np

# ---------------------------------------------------------------------------
# Constants
# ---------------------------------------------------------------------------
MB = 1024 * 1024
VECTOR_BYTES = 64 * MB                      # 64 MB per vector
DTYPE = np.float32                           # 4 bytes / element
ELEMENTS_PER_VECTOR = VECTOR_BYTES // np.dtype(DTYPE).itemsize   # 16 777 216
NUM_PAIRS = 8                                # how many (a, b) pairs to stream


# ---------------------------------------------------------------------------
# 1. Streaming source
# ---------------------------------------------------------------------------
def stream_vectors(num_pairs: int = NUM_PAIRS):
    """
    Yield (index, a, b) tuples where a and b are 64 MB float32 vectors.

    The generator materialises only one pair at a time so peak memory is
    bounded to ~3 × 64 MB = 192 MB (a, b, and the result buffer).
    """
    rng = np.random.default_rng(seed=42)
    for i in range(num_pairs):
        a = rng.random(ELEMENTS_PER_VECTOR, dtype=DTYPE)
        b = rng.random(ELEMENTS_PER_VECTOR, dtype=DTYPE)
        yield i, a, b


# ---------------------------------------------------------------------------
# 2. Timed add — isolates *just* the addition
# ---------------------------------------------------------------------------
def timed_add(a: np.ndarray, b: np.ndarray) -> tuple[np.ndarray, float]:
    """
    Add two vectors and return (result, elapsed_seconds).

    The timer wraps only the np.add call so I/O, allocation, and everything
    else is excluded from the measurement.
    """
    t0 = time.perf_counter()
    result = np.add(a, b)          # ← the only line being timed
    elapsed = time.perf_counter() - t0
    return result, elapsed


# ---------------------------------------------------------------------------
# 3. Runtime estimation helpers
# ---------------------------------------------------------------------------
def estimate_remaining(elapsed_so_far: list[float], pairs_left: int) -> float:
    """
    Return a running-average estimate of remaining add time (seconds).

    Uses the mean of all observed add durations so the estimate stabilises
    as more samples arrive.
    """
    if not elapsed_so_far:
        return float("nan")
    mean_per_pair = sum(elapsed_so_far) / len(elapsed_so_far)
    return mean_per_pair * pairs_left


def throughput_gbs(elapsed_sec: float) -> float:
    """Return add throughput in GB/s (two input vectors read + one output written)."""
    bytes_touched = 3 * VECTOR_BYTES        # read a, read b, write result
    return (bytes_touched / elapsed_sec) / (1024 ** 3)


# ---------------------------------------------------------------------------
# 4. Main
# ---------------------------------------------------------------------------
def main():
    print("=" * 62)
    print(f"  Stream-add benchmark")
    print(f"  Vector size : {VECTOR_BYTES / MB:.0f} MB  ({ELEMENTS_PER_VECTOR:,} float32 elements)")
    print(f"  Pairs       : {NUM_PAIRS}")
    print("=" * 62)

    add_times: list[float] = []

    for idx, a, b in stream_vectors(NUM_PAIRS):
        # ── just the add ────────────────────────────────────────────────
        result, elapsed = timed_add(a, b)
        # ────────────────────────────────────────────────────────────────

        add_times.append(elapsed)
        pairs_done = idx + 1
        pairs_left = NUM_PAIRS - pairs_done
        est_remaining = estimate_remaining(add_times, pairs_left)

        print(
            f"  pair {pairs_done:>2}/{NUM_PAIRS} │ "
            f"add = {elapsed * 1000:7.3f} ms │ "
            f"{throughput_gbs(elapsed):5.2f} GB/s │ "
            f"est. remaining add time = "
            + (f"{est_remaining * 1000:.1f} ms" if pairs_left else "done")
        )

        # Discard result immediately — we are benchmarking, not accumulating
        del result

    # ── Summary ─────────────────────────────────────────────────────────
    total   = sum(add_times)
    mean    = total / len(add_times)
    minimum = min(add_times)
    maximum = max(add_times)

    print("-" * 62)
    print(f"  Total add time : {total  * 1000:8.3f} ms")
    print(f"  Mean  add time : {mean   * 1000:8.3f} ms  ({throughput_gbs(mean):.2f} GB/s avg)")
    print(f"  Min   add time : {minimum* 1000:8.3f} ms")
    print(f"  Max   add time : {maximum* 1000:8.3f} ms")
    print("=" * 62)


if __name__ == "__main__":
    main()
